### PR TITLE
Eavesdropping radio channels that you don't have access to will be identified as "Unknown" rather than its actual name

### LINKED ISSUE #7996: Radio headsets can be heard by all mobs sharing the tile
## Reproduction:

1) Have a headset with a secure channel
2) Be wearing the headset (nobody can hear it if it isn't on someone's ear)
3) Receive sensitive information on a private channel
4) There's an assistant laying on the ground staring at your crotch. They just heard your headset receive that sensitive message

Headsets only send messages to the tile they're on if *someone* is wearing it, so there is no way this is intended eavesdropping behavior.

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
Eavesdropping radio channels that you don't have access to will be identified as "Unknown" rather than its actual name

## Why It's Good For The Game
sometimes you can eavesdrop someone's radio talks but recognising which channel the talks came from is a kind of immersion breaking and giving you metaknowlege.
you will not be capable of automatically recognising the channel names anymore. These will be recognised as Unknown.
This is applied to every radio channel including common/AI private/custom frequencies. If someone is talking through station bounced radio (freq: 144.1), they'll see that as `[144.1] Clownie says, "I love you!"` but you'll see that as `[Unknown] Clownie says, "I love you!"`. If your headset is set to 144.1, you can recognise its frequency.
so, this can hide your own private and secret radio channels... 

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

a doctor (having medical access) eavesdrops someone who has medical/more radio channels. That guy talked to a channel that the doctor doesn't have access to, so it's identified as Unknown.

----------------

image

up: captain
below: CMO

---------------

image
image

someone is holding in 3 hotspot station bounced radios. holder can see the channel names, but someone hearing those can't.

--------------

image
image

radios are placed in the hallway - freq is "common" "144.1" "144.5". the speaker isn't holding those (but common is on their headset), so they can't identify custom freq

--------------

image

It's common channel, but the below one isn't wearing headset, or headset isn't set to common radio channel.

---------------

image
image

they're standing next to intercom. even if they don't have headset or radio, they can identify channel names.

----------------

image

syndi key can let you identify channel names

------------

image

light mode colour

</details>

## Changelog
:cl:
tweak: when you eavesdrop someone's radio talks but you don't have access to those because you don't have an encryption key, you'll fail to identify which channel the talks come from. the channel names will be seen as "Unknown". Being eligible to access to a channel follows these conditions - have a station bounce radio / wear a headset / have a radio implant & antenna mutation / standing next to intercom - and all of them must be turned on.
code: minor change to radio related code (i.e. make defines)
code: set some canhear_range to -1 (i.e. Eminence radio)
fix: fixed a case a mob can't hear their radio when their radio's canhear_range is -1
/:cl:

### CODE REVIEW

File: code/__DEFINES/radio.dm [hunk 1]
```diff
@@ -122,3 +124,8 @@
 #define REQ_DEP_TYPE_ASSISTANCE 	(1<<0)
 #define REQ_DEP_TYPE_SUPPLIES 		(1<<1)
 #define REQ_DEP_TYPE_INFORMATION 	(1<<2)
+
+//range of listening
+#define RADIO_HEARING_RANGE_BASIC_RADIO 3
+#define RADIO_HEARING_RANGE_INTERCOM 2
+#define RADIO_HEARING_RANGE_HEADSET 1
```
Comment (mystery3525) [inline code/__DEFINES/radio.dm]:
```suggestion
#define RADIO_HEARING_RANGE_HEADSET 0
```
I feel as though just general hearing will begin to clutter up if you are, say, standing in a 2 by 2 formation with someone from sec, someone from engineering, etc. We should probably keep this the way it is for headsets.

Comment (EvilDragonfiend) [reply code/__DEFINES/radio.dm]: I really don't know but it's what Ruko requested. I will wait for opinions of how maintainers think.

Comment (PowerfulBacon) [reply code/__DEFINES/radio.dm]: You shouldnt be able to hear someones headset if you are standing next to them, I think only if you are being carried by them or in the same locker as them/on the same tile. Implant radios should never be heard by others.

Comment (PowerfulBacon) [reply code/__DEFINES/radio.dm]: oh wait this is the mic range

Comment (EvilDragonfiend) [reply code/__DEFINES/radio.dm:131]: I have no idea what Bacon is thinking. You triaged Ruko's issue.

Comment (EvilDragonfiend) [reply code/__DEFINES/radio.dm]: I don't think I can fix eavesdropping under someone's crotch, so do you want this to be 0 or what Ruko suggested?

**Comment by EvilDragonfiend** [comment]
Test screenshots uploaded

File: code/__DEFINES/radio.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/__DEFINES/radio.dm]: I think it should be 0 range rather than when standing next to them personally. Implants are not meant to be hearable by other people ever

Comment (EvilDragonfiend) [reply code/__DEFINES/radio.dm]: that's confusing. I didn't touch the hearing range of implants, but did to headsets. revert headsets'?

**Comment by hotimep** [comment]
is syndichat also picked up one tile away by people who have the encryption key on

**Comment by EvilDragonfiend** [comment]
> is syndichat also picked up one tile away by people who have the encryption key on

`add: syndicate encryption key decreases eavesdrop range to 0`

**Comment by hotimep** [comment]
another win for the readers

**Comment by EvilDragonfiend** [comment]
okay, I found out why we hear security radio from nowhere. it was from beepsky that has security radio.
so, this isn't actually a proper way to fix the issue, so I will remove 'closes #7996' part, but it is still good to prevent metainformation.

**Comment by itsmeow** [comment]
Does this affect ghost or admin ears?

**Comment by EvilDragonfiend** [comment]
> Does this affect ghost or admin ears?

ghosts are fine. admin ears are still restricted unless they have a required key.

File: code/__DEFINES/radio.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/__DEFINES/radio.dm]: Revert headsets to a 0 range I think

File: code/modules/mob/living/say.dm [hunk 2]
```diff
@@ -233,6 +233,35 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
+				available_channels += "[R.radio.frequency]"
+				for(var/K in R.radio.keyslot?.channels)
+					available_channels += "[GLOB.radiochannels[K]]"
+		for(var/obj/item/radio/intercom/R in view_or_range(RADIO_HEARING_RANGE_INTERCOM, src, "view"))
+			if(R.listening)
+				available_channels += "[R.frequency]"
+				for(var/K in R.keyslot?.channels)
+					available_channels += "[GLOB.radiochannels[K]]"
+		if(!("[radio_freq]" in available_channels) && !("[FREQ_SYNDICATE]" in available_channels)) // have an enc key or a syndi key, or channel will be seen as unknown
+			radio_freq = FREQ_UNKNOWN
```
Comment (PowerfulBacon) [inline code/modules/mob/living/say.dm]: This code will probably be quite slow, iterating through the entire inventory every time a radio message is heard

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
You'd probably have to cache something somewhere, or determine if the source of Hear() was from another person. (Doesn't hear() only get called once if you heard the original message, since you don't get duplicate chat messages when under someone)

File: code/modules/mob/living/say.dm (hunk 2 above)
Comment (EvilDragonfiend) [reply code/modules/mob/living/say.dm]: good point. the whole code will not be slow, but line 239 seems to be possibly an issue. I will think of a way to cache it.

**Comment by EvilDragonfiend** [comment]
image

radio on their ears

-----

image

radio dropped on their ears / turned off

------

also tested on some cases (like EMP). all things went well.

File: code/modules/mob/living/say.dm [hunk 3]
```diff
@@ -238,6 +238,19 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 		deaf_message = "<span class='notice'>You can't hear yourself!</span>"
 		deaf_type = 2 // Since you should be able to hear yourself without looking
 
+	// this makes you difficult to identify a radio channel as "Unknown" to prevent you to get metaknowledge of radio channel names
+	if(radio_freq)
+		var/radio_found = FALSE
+		if(src.known_channels["[radio_freq]"] || src.known_channels["[FREQ_SYNDICATE]"] || isdead(src))
```
Comment (PowerfulBacon) [inline code/modules/mob/living/say.dm:250]: This will runtime if someone doesn't have known_channels initialised

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Make it so you can't hear radios if you are standing on a tile next to a person, only if you are under them.

**Comment by itsmeow** [comment]
I really like how this looks ingame actually, well made, however I have not looked at the code

File: code/game/objects/items/devices/radio/headset.dm [hunk 4]
```diff
@@ -349,6 +349,7 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 			translate_binary = TRUE
 		if(keyslot2.syndie)
 			syndie = TRUE
+			canhear_range = 0
```
Comment (itsmeow) [inline code/game/objects/items/devices/radio/headset.dm]: Why is this here? Should this also be a define?

File: code/modules/mob/living/say.dm [hunk 5]
```diff
@@ -238,6 +238,19 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
 
+	// this makes you difficult to identify a radio channel as "Unknown" to prevent you to get metaknowledge of radio channel names
+	if(radio_freq && known_channels)
+		var/radio_found = FALSE
+		if(src.known_channels["[radio_freq]"] || src.known_channels["[FREQ_SYNDICATE]"] || isdead(src))
+			radio_found = TRUE
+		if(!radio_found) // standing around intercom can make you understand
+			for(var/obj/item/radio/intercom/R in view_or_range(RADIO_HEARING_RANGE_INTERCOM, src, "view"))
+				if(R.listening)
+					if(R.frequency == "[radio_freq]")
```
Comment (itsmeow) [inline code/modules/mob/living/say.dm]: unnecessary double if statement, use &&

File: code/modules/mob/mob.dm [hunk 6]
```diff
@@ -1271,3 +1271,31 @@
...
+
+/// if you hear a radio talk that you don't have access to for some reason, you can't recognise its channel name. This updates your known channel depending on radios in your possession
+/mob/proc/refresh_known_radio_channels()
+	known_channels = list() // reset first
+	for(var/obj/item/radio/R in src.get_contents())
+		if(R.listening) // if not turned off, you shouldn't be eligible for this.
+			if(istype(R, /obj/item/radio/headset)) // only headset has 2nd keyslot
+				var/mob/living/carbon/C = src
+				var/obj/item/radio/headset/H = R // only headset has `keyslot2` variable
+				if(H == C.ears) // this must be worn by you, not in your inventory.
```
Comment (itsmeow) [inline code/modules/mob/mob.dm]: there are way too many single letter varnames here for this to make any sense

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/radio/headset.dm (hunk 4 above)
Comment (EvilDragonfiend) [reply code/game/objects/items/devices/radio/headset.dm]: eavesdropping syndi channel isn't good even if it's said unknown, so decreasing hearing range forcefully.
also, changed hearing range as a define and to -1. this will even prevent eavesdropping by standing below someone

File: code/game/objects/items/devices/radio/radio.dm [hunk 7]
```diff
@@ -45,6 +45,51 @@
...
+	. = ..()
+	if(current_holder)
+		addtimer(CALLBACK(current_holder, /mob/.proc/refresh_known_radio_channels), 2)
+	current_holder = null
+
+/obj/item/radio/equipped(mob/user, slot)
+	. = ..()
+	current_holder = user
+	if(current_holder)
+		addtimer(CALLBACK(current_holder, /mob/.proc/refresh_known_radio_channels), 2)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/radio/radio.dm]: This 0.2 second delay feels incredibly janky, would it be better to main the list of known radio channels and add new ones when a radio is equipped to the correct slot?

For example, when a security headset is equipped to the ear slot, add 1 to RADIO_CHANNEL_SECURITY.
When a player picks up a security radio add 1 to RADIO_CHANNEL_SECURITY.
When the player takes off their headset remove 1.
If the player drops their security radio, the count is now 0 so they can no longer hear it.

This would also avoid the pretty expensive refresh_known_radio_channels() along with all its snowflake behaviour

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/radio/radio.dm (hunk 7 above)
Comment (EvilDragonfiend) [reply code/game/objects/items/devices/radio/radio.dm]: It is related to how `Destroy()` proc works. It should be called after destroyed, but the game still thinks they can access to the channel, so I made it like that.(I mean, drop/destroyed is called later after a point when I want it to happen) I am doubtful that your suggestion wouldn't change that weird behaviour, and I have no clue how I can fix that without using addtimer.

Comment (EvilDragonfiend) [reply code/game/objects/items/devices/radio/radio.dm]: Well, let me test something then I will tell you

Comment (EvilDragonfiend) [reply code/game/objects/items/devices/radio/radio.dm]: I think changing it to 'add/remove' logic would make it any good. It's because there are not a few things to check
* checks Listening
* checks if it is headset, and be at your ears
* checks if it is in your bag
* checks keyslot
* checks keyslot2
* and so on

and these things are not only single checkable. I mean, these are even should be checked together, which means it needs incredibly a lot of condition checks.
but also how `pickup/dropped/equipped` procs work makes my brain explode.

Comment (PowerfulBacon) [reply code/game/objects/items/devices/radio/radio.dm]: Doesnt destroy call dropped()?

Comment (EvilDragonfiend) [reply code/game/objects/items/devices/radio/radio.dm]: `destory()` calls `dropped()`, but not at the moment I want. That's why I had to put 0.2s delay.

File: code/game/objects/items/devices/radio/radio.dm [hunk 8]
```diff
@@ -70,6 +117,7 @@
 			translate_binary = TRUE
 		if(keyslot.syndie)
 			syndie = TRUE
+			canhear_range = RADIO_HEARING_RANGE_STEALTHY
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/radio/radio.dm]: Syndie radios should only be unhearable if they are in an implant

File: code/modules/mob/mob.dm [hunk 9]
```diff
@@ -1271,3 +1271,34 @@
...
+				if(headset_on_ears == carbon.ears) // this must be worn by you, not in your inventory.
+					known_channels["[headset_on_ears.frequency]"] = TRUE
+					for(var/K in headset_on_ears.keyslot?.channels)
+						known_channels["[GLOB.radiochannels[K]]"] = TRUE
+					for(var/K in headset_on_ears.keyslot2?.channels)
+						known_channels["[GLOB.radiochannels[K]]"] = TRUE
+			else
+				known_channels["[mob_radio.frequency]"] = TRUE
+				for(var/K in mob_radio.keyslot?.channels)
+					known_channels["[GLOB.radiochannels[K]]"] = TRUE
```
Comment (PowerfulBacon) [inline code/modules/mob/mob.dm]: Not sure how difficult it would be to implement, but can I suggest a far, far better approach?
Store a dictionary on the mob with a list of channels that they know and how many times they know them.
When a headset is equipped to the ear slot, increment the counter by 1.
When a headset is unequipped **from the ear slot**, increment the counter by 1.
For radios do the same, but have it when added to any slot on the mob.

This would completely avoid looping over everything on the mob (get_contents() can be real slow) and would make this not require a loop at all. The only challenge is ensuring consistency when decrementing and incrementing

Comment (EvilDragonfiend) [reply code/modules/mob/mob.dm]: image

this is why it's not possible. If you move a thing within in your hand, it calls 'pick()' verb every time.

**Comment by EvilDragonfiend** [comment]
gonna try something
